### PR TITLE
Update caches instead of invalidating them

### PR DESCRIPTION
**Pull request by sadhansood**
## Description 

As title says, we are going to start updating index caches with account balances instead of invalidating them. The motivation is that going to db is expensive for some accounts which hold tens and hundreds thousand keys, and by keep balance updated in cache we see a p99 of less than 0.01 seconds
## Test Plan 

Enhanced existing tests

### CODE REVIEW

**Review by patrickkuo — APPROVED** [review]
LGTM

File: crates/sui-json-rpc/src/coin_api.rs [hunk 1]
```diff
@@ -223,8 +223,10 @@ impl CoinReadApiServer for CoinReadApi {
             })?;
         Ok(Balance {
             coin_type: coin_type.to_string(),
-            coin_object_count: balance.num_coins,
-            total_balance: balance.balance,
+            coin_object_count: usize::try_from(balance.num_coins)
+                .map_err(|e| Error::SuiError(SuiError::ExecutionError(e.to_string())))?,
+            total_balance: u128::try_from(balance.balance)
+                .map_err(|e| Error::SuiError(SuiError::ExecutionError(e.to_string())))?,
```
Comment (longbowlu) [inline crates/sui-json-rpc/src/coin_api.rs:229]: plz do not use execution error, it signals something very different (usually a VM error)

File: crates/sui-json-rpc/src/coin_api.rs [hunk 2]
```diff
@@ -244,18 +246,19 @@ impl CoinReadApiServer for CoinReadApi {
...
-                }
+        let mut result = vec![];
+        for (coin_type, balance) in all_balance.iter() {
+            result.push(Balance {
+                coin_type: coin_type.to_string(),
+                coin_object_count: usize::try_from(balance.num_coins)
+                    .map_err(|e| Error::SuiError(SuiError::ExecutionError(e.to_string())))?,
+                total_balance: u128::try_from(balance.balance)
+                    .map_err(|e| Error::SuiError(SuiError::ExecutionError(e.to_string())))?,
+                // note: LockedCoin is deprecated
```
Comment (longbowlu) [inline crates/sui-json-rpc/src/coin_api.rs:257]: ditto.

The merge-op thing has a lot of integer conversions and I think unwrap may be better because the max balance of our Coin type is u64 so it should't be possible for this to fail. And if it fails, it'd better fail fast and scream

File: crates/sui-storage/src/indexes.rs [hunk 3]
```diff
@@ -295,17 +295,26 @@ impl IndexStore {
         let coin_delete_keys = object_index_changes
             .deleted_owners
             .iter()
+            .dedup()
```
Comment (longbowlu) [inline crates/sui-storage/src/indexes.rs:298]: why is this needed?

File: crates/sui-storage/src/indexes.rs [hunk 4]
```diff
@@ -295,17 +295,26 @@ impl IndexStore {
...
                     )
                 });
-                address_coin_type_to_invalidate.insert((*owner, coin_type_tag.clone()));
-                addresses_to_invalidate.insert(*owner);
+                let map = balance_changes.entry(*owner).or_insert(HashMap::new());
+                let entry = map.entry(coin_type_tag.clone()).or_insert(TotalBalance {
+                    num_coins: 0,
+                    balance: 0
+                });
+                let existing = self.tables.coin_index.get(&(*owner, coin_type_tag.to_string(), *obj_id));
```
Comment (longbowlu) [inline crates/sui-storage/src/indexes.rs]: would there be a data race here?

File: crates/sui-storage/src/indexes.rs [hunk 5]
```diff
@@ -323,7 +332,7 @@ impl IndexStore {
         // due to a successful or failed transaction.
         let coin_add_keys = object_index_changes
         .new_owners
-        .iter()
+        .iter().dedup()
```
Comment (longbowlu) [inline crates/sui-storage/src/indexes.rs:333]: ditto

File: crates/sui-storage/src/indexes.rs [hunk 6]
```diff
@@ -339,8 +346,23 @@ impl IndexStore {
...
             });
-            address_coin_type_to_invalidate.insert((*owner, coin_type_tag.clone()));
-            addresses_to_invalidate.insert(*owner);
+            let map = balance_changes.entry(*owner).or_insert(HashMap::new());
+            let entry = map.entry(coin_type_tag.clone()).or_insert(TotalBalance {
+                num_coins: 0,
+                balance: 0
+            });
+
+            let existing = self.tables.coin_index.get(&(*owner, coin_type_tag.to_string(), *obj_id));
```
Comment (longbowlu) [inline crates/sui-storage/src/indexes.rs:355]: would there be a data race here

File: crates/sui-json-rpc/src/coin_api.rs (hunk 1 above)
Comment (sadhansood) [reply crates/sui-json-rpc/src/coin_api.rs]: done

File: crates/sui-json-rpc/src/coin_api.rs (hunk 2 above)
Comment (sadhansood) [reply crates/sui-json-rpc/src/coin_api.rs]: sure, done

File: crates/sui-storage/src/indexes.rs (hunk 3 above)
Comment (sadhansood) [reply crates/sui-storage/src/indexes.rs]: my bad, it isn't removed

File: crates/sui-storage/src/indexes.rs (hunk 4 above)
Comment (sadhansood) [reply crates/sui-storage/src/indexes.rs]: There could be, although I am not 100% sure because it seems like input and output objects of one transaction cannot intersect with another txns for owned objects but not for shared objects. I added a lock now which prevents any possible race b/w concurrent coin index access for an address. This lock is held very briefly.

File: crates/sui-storage/src/indexes.rs (hunk 5 above)
Comment (sadhansood) [reply crates/sui-storage/src/indexes.rs]: removed

File: crates/sui-storage/src/indexes.rs (hunk 6 above)
Comment (sadhansood) [reply crates/sui-storage/src/indexes.rs]: ditto

**Comment by sadhansood** [comment]
Addressed @longbowlu comments
